File: classes/201-f25/Comparing-Python-and-Kotlin/die.py
"""
   die.py
   Jeff Ondich, 2013-01-05
   Anna Rafferty

   Modified by Tanya Amert for Spring 2025

   A simple class representing an n-sided die (as used in various board games).
   
   Intended as the Python half of parallel examples in Python and
   Kotlin. See Die.kt.
"""

import random

class Die:
    def __init__(self, numberOfSides):
        self.numberOfSides = numberOfSides
        self.currentValue = 1

    def roll(self):
        self.currentValue = random.randint(1, self.numberOfSides)

    def getValue(self):
        return self.currentValue

    def setValue(self, value):
        self.valcurrentValueue = value

# This if-statement ensures that this code will only be run when die.py is
# executed as a program in its own right.  If a different module does
# "import die", the if-condition will be false.
if __name__ == '__main__':
    # Let's play Yahtzee
    dice = []
    dice.append(Die(6))
    dice.append(Die(6))
    dice.append(Die(6))
    dice.append(Die(6))
    dice.append(Die(6))

    for die in dice:
        die.roll()

    total = 0
    for die in dice:
        print(die.getValue(), end=" ")
        total += die.getValue()
    print()
    print("Total:", total)

    # This is really long condition on multiple lines
    if dice[0].getValue() == dice[1].getValue() \
           and dice[1].getValue() == dice[2].getValue() \
           and dice[2].getValue() == dice[3].getValue() \
           and dice[3].getValue() == dice[4].getValue():
        print("Yahtzee!")

File: classes/201-f25/Comparing-Python-and-Kotlin/Die.kt
/*
 * Die.kt
 * Tanya Amert, 2025-03-31
 * Based on Die.java by Jeff Ondich, Anna Rafferty
 *
 * A simple class representing an n-sided die (as used in various board games).
 * 
 * After running this class, make a variable that represents a 9-sided die, and
 * then print out the result of rolling it.  Which lines in the file represent the 
 * constructor for the die object?
 *
 * This is the Kotlin half of a pair of parallel examples in Python and Kotlin.
 * See die.py.
 */

import kotlin.random.Random

class Die(val n: Int) {
    // These are the instance variables (like the "self.*" stuff in Python)
    val numberOfSides: Int = n // val: always has value n
    var currentValue: Int = 1  // var: its value will vary
    
    fun roll(): Unit { // returns nothing
        currentValue = Random.nextInt(1, numberOfSides+1) // exclusive end of range
    }

    fun getValue(): Int {
        return currentValue
    }

    fun setValue(newValue: Int): Unit {
        currentValue = newValue
    }
}

fun main() {
    // Let's play Yahtzee
    val dice: MutableList<Die> = mutableListOf()
    dice.add(Die(6))
    dice.add(Die(6))
    dice.add(Die(6))
    dice.add(Die(6))
    dice.add(Die(6))

    for (die: Die in dice) {
        die.roll()
    }

    var total: Int = 0
    for (die: Die in dice) {
        print("${die.getValue()} ")
        total += die.getValue()
    }
    println("\nTotal: $total")

    if (dice[0].getValue() == dice[1].getValue() &&
        dice[1].getValue() == dice[2].getValue() &&
        dice[2].getValue() == dice[3].getValue() &&
        dice[3].getValue() == dice[4].getValue()) {
        println("Yahtzee!");
    }
}